### PR TITLE
Implement neighbor root polling

### PR DESCRIPTION
**Pull request by dharjeezy**
make use of default  zero root hashes

### CODE REVIEW

File: primitives/src/traits/merkle_tree.rs [hunk 1]
```diff
@@ -7,6 +7,8 @@ pub trait TreeInterface<AccountId, TreeId, Element> {
 	fn create(creator: Option<AccountId>, depth: u8) -> Result<TreeId, dispatch::DispatchError>;
 	/// Adds members/leaves to the tree
 	fn insert_in_order(id: TreeId, leaf: Element) -> Result<Element, dispatch::DispatchError>;
+
+	fn zero_root(i: u8) -> Result<[u8; 32], dispatch::DispatchError>;
```
Comment (drewstone) [inline primitives/src/traits/merkle_tree.rs:11]: Can we add documentation to this?

File: primitives/src/traits/linkable_tree.rs [hunk 2]
```diff
@@ -23,6 +26,7 @@ pub trait LinkableTreeInterface<C: LinkableTreeConfig> {
 		id: C::TreeId,
 		leaf: C::Element,
 	) -> Result<C::Element, dispatch::DispatchError>;
+	fn zero_root(i: u8) -> Result<[u8; 32], dispatch::DispatchError>;
```
Comment (drewstone) [inline primitives/src/traits/linkable_tree.rs:29]: Documentation, describe what the zero root is.

**Review by drewstone — CHANGES_REQUESTED** [review]
LGTM! Let's move the default hashes into the chain spec / genesis configuration. I think that would be an improvement greatly needed.

File: pallets/linkable-tree/src/lib.rs [hunk 3]
```diff
@@ -90,6 +90,8 @@ pub use weights::WeightInfo;
 
 pub use pallet::*;
 
+pub type DefaultMerkleRootIndexOf<T, I> = <T as pallet_mt::Config<I>>::DefaultMerkleRootIndex;
```
Comment (dutterbutter) [inline pallets/linkable-tree/src/lib.rs:93]: Please add a brief definition to public type.

File: pallets/linkable-tree/src/lib.rs [hunk 4]
```diff
@@ -392,12 +403,51 @@ impl<T: Config<I>, I: 'static> LinkableTreeInspector<LinkableTreeConfigration<T,
...
+			// and push them to the new_edges list
+			let edges = EdgeList::<T, I>::iter_prefix(id)
+				.into_iter()
+				.map(|x| new_edges.push(x.1))
+				.collect::<Vec<_>>();
+
+			// get the max_edges
+			let max_edges = Self::max_edges(id).try_into().unwrap();
+
+			if edges.len() <= max_edges {
```
Comment (drewstone) [inline pallets/linkable-tree/src/lib.rs:419]: Should be <, no?

File: pallets/mt/src/lib.rs [hunk 5]
```diff
@@ -252,18 +263,29 @@ pub mod pallet {
...
+		pub default_zero_root_hashes: Option<Vec<(u8, [u8; 32])>>,
 	}
 
 	#[cfg(feature = "std")]
 	impl<T: Config<I>, I: 'static> Default for GenesisConfig<T, I> {
 		fn default() -> Self {
-			Self { phantom: Default::default(), default_hashes: None }
+			Self {
+				phantom: Default::default(),
+				default_hashes: None,
```
Comment (drewstone) [inline pallets/mt/src/lib.rs:274]: Redundant

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks good! I would remove usage of default hashes since that is what was previously used. We shouldn't need it anymore assuming you are relying on this new collection. Therefore, we'd want to remove wherever it's called/creates/queries and replace with the new one.

**Pull request closed without merging**